numpy/core/src/multiarray/datetime.c

```diff
@@ -430,6 +431,104 @@ npy_int64 extract_unit(npy_datetime *d, npy_datetime unit) {
     return div;
 }
 
+/*
+ * Convert datetime64 to a tuple (quad_centuries, microseconds, attoseconds)
+ * then hash the tuple.
+ * quad_centuries, because it's easy to convert both (number_of_years) and
+ * (number_of_days) to (number_of_400_year_periods).
+ * microseconds, because 400 years of microseconds fits into 54 bits.
+ * attoseconds, because that's the highest resolution.
+ */
+NPY_NO_EXPORT npy_int64
+datetime_hash(PyArray_DatetimeMetaData *meta, npy_datetime dt)
+{
+    npy_datetimestruct dts;
+    npy_int64 quad_centuries;
+    npy_int64 microseconds;
+    npy_int64 attoseconds;
+
+    dts.year = 1970;
+    dts.month = 1;
+    dts.day = 1;
+
+    dt *= meta->num;
```

Comment:
There's a risk of overflow here. Not sure if we handle that in the datetimestruct - if not, then a to-do comment is probably sufficient.

Reply:
There is a to-do comment in convert_datetime_to_datetimestruct, so I put the same comment in here too.